File: Algorithms/C++/MergeSort.cpp
#include <iostream>
#include <vector>

using namespace std; 

template <typename Type>
void displayVector(vector<Type>& vec) {
    for (int i = 0; i < vec.size(); i++) {
        cout << vec.at(i) << " ";
    }
    cout << endl;
}

template <typename Type>
void merge(vector<Type>& vec, int left, int right, int middle) {
    vector<Type> tempVec;
    
    int i = left; // starting index for left half
    int j = middle + 1; // starting index for right half

    // comparing and copying with the indexes
    while (i <= middle && j <= right) {
        if(vec[i] <= vec[j]) {
            tempVec.push_back(vec[i]);
            i++;
        }
        else {
            tempVec.push_back(vec[j]);
            j++;
        }
    }

    // now we have to make sure all data is inserted
    while (i <= middle) {
        tempVec.push_back(vec[i]);
        i++;
    }
    while (j <= right) {
        tempVec.push_back(vec[j]);
        j++;
    }

    // now copy over elements
    // recursion is why we used the left starting index
    // if we had just started from index 0 instead we would overwrite elements not belonging to the subarray we are currently in 
    for(int z = 0; z < tempVec.size(); z++) {
        vec[left + z] = tempVec[z];
    }
}

template <typename Type>
void mergeSort(vector<Type>& vec, int left, int right) {
    // this is where we are going to break up everything with recursion

    // if we have reached the smallest possible list, left index is greater than right
    if (left >= right) {
        return;
    }
    int middle = left + ((right - left) / 2);

    // this below splits up the the vector into smaller lists each time for both the left and right sides
    mergeSort(vec, left, middle);
    mergeSort(vec, middle + 1, right);

    merge(vec, left, right, middle);
}



int main() {
    // later make it where you can enter your input as well in a seperate option.

    cout << "\nWelcome to the Merge Sort Algorithm (C++), created by Zane Christe\n";
    vector<int> vec1 = {38, 27, 43, 3, 9, 82, 10, 7}; // size is 8
    
    cout << "List before sorting: ";
    displayVector(vec1);

    mergeSort(vec1, 0, vec1.size() - 1);

    cout << "\nList after sorting: ";
    displayVector(vec1);

    return 0;      
}

/*
    cd Algorithms/C++
    g++ -std=c++11 MergeSort.cpp -o tempExecutable && ./tempExecutable && rm tempExecutable

*/

File: Algorithms/Python/MergeSort.py
# just run python file with upper right corner and press "Run Python File" from drop down menu
# What we are making is a mergesort algorithm in python, it is a divide and conquer algo so we want to use recursion
# remember it divides the arrays in half and sorts recursively and then merges the sorted halves

# The helper function of the mergesort
def merge(left, right): 
    results = []
    i = 0
    j = 0
    while i < len(left) and j < len(right):
        # compare each element to see what order it needs to be in
        if left[i] < right[j]:
            results.append(left[i])
            i = i + 1
        else:
            results.append(right[j])
            j = j + 1

    # Append the remaining elements from the left and right arrays
    results.extend(left[i:])
    results.extend(right[j:])

    return results
    

def mergeSort(array):
    if len(array) <= 1:
        return array
    
    # find the middle index
    middle = len(array) //  2

    # using slicing notation to access parts of the list
    leftHalf = array[:middle] # slice of the list up to but not including the middle index
    rightHalf = array[middle:] # slice of the list including the middle index to the end
    
    # using recursion to keep making smaller lists
    sortedLeft = mergeSort(leftHalf)
    sortedRight = mergeSort(rightHalf)

    return merge(sortedLeft, sortedRight)
    
    

def main():
    # later make it where you can enter your input as well in a seperate option.

    print("\nWelcome to the Merge Sort Algorithym, created by Zane Christe\n")
    intList = [7, 3, -5, 18, 1000, 1522, -9, 1, 9, 420, 69, 777, -252, -6, 4, 99, 0]
    print("List before sorting: ")
    print(intList)

    print("\nList after sorting: ")
    sortedList = mergeSort(intList)
    print(sortedList)

main()